### DISCUSSION TITLE
unityApp.send() dosen't get invoked when json is passed after update from Redux store

### DISCUSSION THREAD

**Discussion by swayam1998** [jeffreylanters/react-unity-webgl#199] [Help Wanted!] [answered]
So I have been trying to communicate with Unity webgl from my react app, using JSON. Here connect() is invoked when a button is clicked. But unityApp.send() doesn't get invoked when I use planet?.port while setting the port in JSON.stringify (There are no Debug.log() outputs from the webgl ). Strangely it gets invoked when raw port number (eg 7000) is passed and everything works normally.

```
var planets = useSelector((state) => state.planets)

...

 const connect = () => {
        const planet = planets[0];
        console.log("Connecting to :", planet?.port, "of type: ", typeof(planet?.port));
        const jsonData = JSON.stringify({
            "networkAddress": "localhost",
            "port": planet?.port
        });
        unityApp.send('NetworkManager', 'ConnnectToAddress', jsonData);
    };
``` 

I am stuck and am unable to figure out what the problem is. I am using Unity 2019 and react-unity-webgl 7.1.10 .

**Comment by jeffreylanters** [comment, marked as answer]
Hi. If I understand you correctly, the send method does not seem to get invoked when you optionally unwrap `port` from `planet`, but does work when you set the `port` to a hard coded value? Can you console log and share both results of `jsonData` with me?

**Reply by swayam1998** [reply]
I just observed this now, unityApp.send() seems to work before I make the api call and update the redux store. After I update it, it stops working, even if I use the hard coded value. 
![image](https://user-images.githubusercontent.com/31216904/120244746-b69cc400-c288-11eb-8c6f-80c0f069eecb.png)
Before the api call and update.

![image](https://user-images.githubusercontent.com/31216904/120244821-e1871800-c288-11eb-9ecd-3f6ab7af29d6.png)
After the api call and update. 
I console logged the unityApp object just to be sure

**Reply by swayam1998** [reply]
I think I've found my mistake, when the redux store got updated the component with unityApp got updated (the uniqueID changed as seen above). So I used the code below to instanciate the unityApp, which  stopped it from reloading and it seems to work.
```
const unityApp = useMemo(() => new UnityContent(
        viewsJSON, loader
    ), [viewsJSON, loader]);
```
But I am still unsure why unityApp.send() wont work if it got updated. useMemo is better performance wise too, but I will still try to find out.

**Reply by jeffreylanters** [reply]
Hi. Yes, since you're using hooks, you component will rerender each time a state changes. Which will result in the creation of a new UnityContext object each time. To prevent this, I recommend creating the Unity Context outside of your component as show in the examples.

```js
import React from "react";
import Unity, { UnityContext } from "react-unity-webgl";

const unityContext = new UnityContext({
  loaderUrl: "build/myunityapp.loader.js",
  dataUrl: "build/myunityapp.data",
  frameworkUrl: "build/myunityapp.framework.js",
  codeUrl: "build/myunityapp.wasm",
});

function App() {
  function connect() {
    var jsonData = JSON.stringify({});
    unityContext.send("NetworkManager", "ConnnectToAddress", jsonData);
  }

  return (
    <div>
      <button onClick={connect}>Connect</button>
      <Unity unityContext={unityContext} />
    </div>
  );
}
```

I hope this helps you further, best of luck with your project!

**Reply by swayam1998** [reply]
Thanks. I was over-complicating things, with my previous code.

**Reply by jeffreylanters** [reply]
No worries, glad you've figured it out. Best of luck with your project!

**Reply by Nipunc55** [reply]
Thanks @jeffreylanters
